### PR TITLE
tests: updata the rbtree testcase

### PR DESCRIPTION
**Pull request by Zhaoningx**
Add a new testcase to:
        Verify some operations of rbree are running in logarithmic time.
        Verify an user defined structure contains rbtree node works.
        Verify "for each"style APIs work.

Notice: I changed the directory tests/unit to tests/benchmarks/.  Do you think this change is OK?

Signed-off-by: Ningx Zhao <ningx.zhao@intel.com>

### CODE REVIEW

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (andyross) [inline tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:78]: I'm completely confused about what's happening here.  Are you trying to shuffle the integers between 0 and 511 into the random_num array such that they appear in a random order, but only once each?  If so, I don't think this code works correctly.  It's also needlessly O(N^2) in the size of the array.  Maybe try this instead (untested, but the basic idea)?:

```c


void shuffle_random(void)
{
    int n = ARRAY_SIZE(random_num);

    /* Populate values */
    for (int i = 0; i < n; i++) {
        random_num[i] = i;
    }

    /* Randomize each by swapping with one of the remaining ones */
    for (int i = 0; i < n; i++) {
        int j = next_rand_mod(n - i);
        int tmp = random_num[i];

        random_num[i] = random_num[j];
        random_num[j] = tmp;
    }
}
```

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (andyross) [inline tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:128]: "side", not size

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (andyross) [inline tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:148]: General nitpick: a bug in the test logic or the tree will show up here as an infinite loop.  That **is** a test failure, but it's a very expensive one.  Though in practice this code and the test aren't actually buggy.

File: tests/unit/list/dlist.c
Comment (andyross) [inline tests/unit/list/dlist.c:170]: Please add new misspellings in a separate patch.

**Review by andyross — DISMISSED** [review]
Some nitpicks, and a randomizer that I think needs to be fixed.  Broadly I like the design here very much, checking the tree height instead of clock times.

File: lib/os/rb.c
Comment (Zhaoningx) [reply lib/os/rb.c]: OK, I have reverted it.

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:73]: Yep, this code just copied from the existing rbtree test. Maybe it needn't do this, as you see, these random numbers are used as a index of the rbnode array. I think it's more representative to insert a tree by disorderly ordinal data than orderly ordinal data. And, the testcase must be compatible with all platforms, the API sys_rand32_get() is not effective for all. So maybe I changed the method that inserting a tree, just using some orderly ordinal data. Do you think the method is OK?

File: tests/unit/list/dlist.c
Comment (Zhaoningx) [reply tests/unit/list/dlist.c]: OK, I will do this in a separate patch.

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:154]: I met this condition before. When I run the code, it will do as an infinite loop, until the stack broken. I think the reason is that inserting two same nodes into the tree, it will not work normally. So do you think how can I get the height of node in the tree instead this.:)

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c]: The code is very good, but there are a few same numbers,as you know, it can't insert a tree using two same data.  If I need to use disorderly ordinal data, I think I can use this code instead and fix it.

**Comment by Zhaoningx** [comment]
Hi @andyross , thank you very much for your comments, I have changed my code, I think maybe I needn't disorder numbers, and I used the method of recurse to search height of the node of a tree. Then the height of the comparison node is less than or equal to the worst-case height of the equilibrium tree.Do you think this is effective?Adding some comments to point some issues if there are.Thanks again.

**Comment by Zhaoningx** [comment]
Hi @andyross @andrewboie @nashif , would you like to review this PR again, thanks.

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (LeiW000) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:51]: I agree with Andy. You don't need to zero test_tree_l.

Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:51]: Hi @andyross @LeiW000 ,Thanks for your comments, after I double check this code, the 'test_tree_l' and 'node' is local variable and this sub-function just is called once time, so I removed "static",and zero-initialized manually. :)

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c]: I changed another method to implement that finding the height of the tree.

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c]: OK, Thanks.

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c]: Done, Thanks

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c]: Done, thanks.

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (LeiW000) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:51]: Sorry, I was wrong.

**Review by LeiW000 — APPROVED** [review]
It looks good to me.

**Review by tyrantlizard77 — COMMENTED** [review]
Nice idea on using tree height to demonstrate O(log n)!

This generally looks good, please address recent feedback from others.

**Comment by Zhaoningx** [comment]
> Nice idea on using tree height to demonstrate O(log n)!
> 
> This generally looks good, please address recent feedback from others.

Hi @andrewboie ,I think I have fixed all feedback about @andyross and @LeiW000 .

**Comment by LeiW000** [comment]
@andyross , @andrewboie , could you take look at the PR again? Please approve it if you have no other comment.

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:47]: Okey, I have fixed this code.

**Comment by Zhaoningx** [comment]
Hey @andyross  and @andrewboie , I have fixed all feedbacks, would you please review this PR, waiting for more comments.

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (ceolin) [inline tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:83]: I would change this algorithm, you don't need to rely on a global variable.
Just add this parameter to the function to keep the height track, something like:
```
int search_height_recurse(struct rbnode *node, struct rbnode *final_node, uint32_t current_height)
{
	if (node == NULL) {
		return -1;
	}

	if (node == final_node) {
		return current_height;
 	}

	current_height++;
	struct rbnode *ch = z_rb_child(node, tree.lessthan_fn(finial_node, node));
	return search_height_recurse(ch, final_node, current_height);
}
```

Note, I haven't check the algorithm above accuracy, just the idea. Then you don't need to keep that global, you just call this function with initial value 0.

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (ceolin) [inline tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:87]: s/finial_node/final_node/

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (ceolin) [inline tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:106]: Make this value a constant close to TREE_SIZE definition, so if one needs to change that value will easily see that this value has to be changed as well.

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (ceolin) [inline tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:109]: Please look my suggestion about this function. If you want to keep the current implementation there is not need to that function return an int if the result is ignored.

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (ceolin) [inline tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:118]: typos

s/maxmum/maximum/
s/minmum/minimum/

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (ceolin) [inline tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:119]: The period is making this sentence difficult to read

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (ceolin) [inline tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:131]: nitpick, why are you initializing them ? Aren't they static global variables not used before ?

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (ceolin) [inline tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:49]: Do these variables need to be static ?

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (andrewboie) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:131]: agreed these are in BSS should already be zeroed

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:105]: Yup, you are right,I also think it's not flexible, but it's hard to float point arithmetic in zephyr kernel.

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c]: Good to know, I have removed 'memset'.

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c]: Double check it and I think it's local variables and do not need to return, so I don’t think need to add 'static'. And I have removed it, thanks.

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c]: The code looks great, and I do a few modifications to make it can work better. I have used the code,thanks.

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c]: OK, I defined it below the macro TREE_SIZE.  BTW, Thanks for your all suggestions about this PR, if there are not anything improperly, would you approve the PR?

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (ceolin) [inline tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:14]: `static` to limit this variable scope

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (ceolin) [inline tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:81]: previous function you use an unsigned int in the for loop, here you are using int, please make it consistent.

**Review by ceolin — COMMENTED** [review]
Some cosmetic issues, other than that looks great

File: tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c
Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c]: I totally agree with you, thanks.

**Review by ceolin — APPROVED** [review]

**Comment by Zhaoningx** [comment]
Hi @andyross @andrewboie @nashif , would you take a look of this PR, the PR has got two approvals, if there are not anything wrong, could you merge it, thanks.

**Review by tyrantlizard77 — APPROVED** [review]
LGTM.
@andyross wrote the rbtree code so will need his blessing.

**Comment by Zhaoningx** [comment]
@andyross would you like to just spend a few time to review this PR?

**Comment by Zhaoningx** [comment]
Hi @andyross , Could you take a look at this PR? Thanks!

File: tests/benchmarks/data_structure_perf/rbtree_perf/testcase.yaml
Comment (nashif) [inline tests/benchmarks/data_structure_perf/rbtree_perf/testcase.yaml:3]: not sure what req-test is, please do not add arbitrary tags that are difficult to understand

Comment (Zhaoningx) [reply tests/benchmarks/data_structure_perf/rbtree_perf/testcase.yaml:3]: Hi @nashif , I have re-pushed the PR to modify the tags and make them more clear. PR link: https://github.com/zephyrproject-rtos/zephyr/pull/28110

**Pull request merged**
